server/views.py

```diff
@@ -1,444 +1,331 @@
-    if not m:
-        raise ValidationError('Enter a Google Sheets URL')
-    spreadsheet_id = m.group(1)
-    http = google_oauth.http()
-    discoveryUrl = ('https://sheets.googleapis.com/$discovery/rest?'
-                    'version=v4')
-    service = discovery.build('sheets', 'v4', http=http,
-                              discoveryServiceUrl=discoveryUrl)
-
-    try:
-        result = service.spreadsheets().values().get(
-            spreadsheetId=spreadsheet_id, range=sheet_range).execute()
-    except errors.HttpError as e:
-        raise ValidationError(e._get_reason())
-    values = result.get('values', [])
-
-    if not values:
-        raise ValidationError('Sheet is empty')
-    headers = [h.lower() for h in values[0]]
-    rows = [
-        {k: v for k, v in itertools.zip_longest(headers, row, fillvalue='')}
-        for row in values[1:]
-    ]
-    if len(set(headers)) != len(headers):
-        raise ValidationError('Headers must be unique')
-    elif not all(re.match(r'[a-z0-9]+', h) for h in headers):
-        raise ValidationError('Headers must consist of digits and numbers')
-    return headers, rows
-
-def validate_room(exam, room_form):
-    room = Room(
-        exam_id=exam.id,
-        name=slug(room_form.display_name.data),
-        display_name=room_form.display_name.data,
-    )
-    existing_room = Room.query.filter_by(exam_id=exam.id, name=room.name).first()
-    if existing_room:
-        raise ValidationError('A room with that name already exists')
-    headers, rows = read_csv(room_form.sheet_url.data, room_form.sheet_range.data)
-    if 'row' not in headers:
-        raise ValidationError('Missing "row" column')
-    elif 'seat' not in headers:
-        raise ValidationError('Missing "seat" column')
-
-    x, y = 0, -1
-    last_row = None
-    for row in rows:
-        seat = Seat()
-        seat.row = row.pop('row')
-        seat.seat = row.pop('seat')
-        seat.name = seat.row + seat.seat
-        if not seat.name:
-            continue
-        if seat.row != last_row:
-            x = 0
-            y += 1
-        else:
-            x += 1
-        last_row = seat.row
-        x_override = row.pop('x', None)
-        y_override = row.pop('y', None)
+from server.models import db, Exam, Room, Seat, Student
+from server.form import ExamForm, RoomForm, ChooseRoomForm, ImportStudentFromSheetForm, \
+    ImportStudentFromCanvasRosterForm, DeleteStudentForm, AssignForm, EmailForm
+from server.utils.auth import google_oauth
+import server.utils.canvas as canvas_client
+from server.utils.data import parse_form_and_validate_room, validate_students, \
+    parse_student_sheet, parse_canvas_student_roster
+from server.utils.exception import DataValidationError
+from server.utils.url import apply_converter
+from server.utils.assign import assign_students
+from server.utils.email import email_students
+
+apply_converter()
+
+# region Offering CRUDI
+
+
+@app.route('/')
```

Comment:
yeah we should have some controllers directory

Reply:
yes